charmcraft/linters.py

```diff
@@ -0,0 +1,180 @@
+            with dispatch.open("rt", encoding="utf8") as fh:
+                last_line = None
+                for line in fh:
+                    if line.strip():
+                        last_line = line
+                if last_line:
+                    entrypoint_str = shlex.split(last_line)[-1]
+        except (IOError, UnicodeDecodeError):
+            return cls.Result.unknown
+
+        entrypoint = basedir / entrypoint_str
+        if entrypoint.suffix == ".py" and os.access(entrypoint, os.X_OK):
+            return cls.Result.python
+        return cls.Result.unknown
+
+
+class Framework:
+    """Check on which framework the charm is based on.
+
+    Currently it detects if the Operator Framework is used, if...
+
+    - the language trait is set to python
+    - the charm contains venv/ops
+    - the charm imports ops in the entry point.
+
+    ...or the Reactive Framework is used, if the charm...
+
+    - has a metadata.yaml with "name" in it
+    - has a reactive/<name>,py file that imports "charms.reactive"
+    - has a file inside "wheelhouse" dir whose name starts with "charms.reactive-"
+    """
+
+    check_type = CheckType.trait
+    name = "framework"
+    url = "https://juju.is/docs/sdk/charmcraft-analyze#heading--framework"
+    text = "The charm is based on the Operator Framework."
+
+    # different result constants
+    Result = namedtuple("Result", "operator reactive unknown")(
+        operator="operator", reactive="reactive", unknown=UNKNOWN
+    )
+
+    @staticmethod
+    def _get_imports(filepath: pathlib.Path) -> Generator[List[str], None, None]:
+        """Parse a Python filepath and yield its imports.
+
+        If the file does not exist or can't be parsed, return empty. Otherwise
+        return the name for each imported module, splitted by possible dots.
+        """
+        if not os.access(filepath, os.R_OK):
+            return
+        try:
+            parsed = ast.parse(filepath.read_bytes())
+        except SyntaxError:
+            return
+
+        for node in ast.walk(parsed):
+            if isinstance(node, ast.Import):
+                for name in node.names:
+                    yield name.name.split(".")
+            elif isinstance(node, ast.ImportFrom):
+                yield node.module.split(".")
+
+    @classmethod
+    def _check_operator(cls, basedir: pathlib.Path) -> bool:
+        """Detect if the Operator Framework is used."""
+        language_info = shared_state[Language.name]
+        if language_info["result"] != Language.Result.python:
+            return False
+
+        opsdir = basedir / "venv" / "ops"
+        if not opsdir.exists() or not opsdir.is_dir():
+            return False
+
+        entrypoint = language_info["entrypoint"]
+        for import_parts in cls._get_imports(entrypoint):
+            if import_parts[0] == "ops":
+                return True
+        return False
+
+    @classmethod
+    def _check_reactive(cls, basedir: pathlib.Path) -> bool:
```

Comment:
what i interesting here is that this should also be Language.Result.python and we might need to redo our python check

Reply:
Let's go back to the drawing board?

Reply:
Ok, we're not changing the Language result for this case.